kernel/aster-nix/src/fs/utils/flock.rs

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::fmt;
+use core::ptr;
+
+use ostd::sync::WaitQueue;
+
+use crate::{
+    fs::{file_handle::FileLike, inode_handle::InodeHandle},
+    prelude::*,
+};
+
+/// Kernel representation of FLOCK
+pub struct Flock {
+    /// Owner of FLOCK, an opened file descriptor holding the lock
+    owner: Weak<dyn FileLike>,
+    /// Type of lock, SH_LOCK or EX_LOCK
+    type_: FlockType,
+    /// Optional waiters that are blocking by the lock
+    waitqueue: Option<WaitQueue>,
```

Comment:
See the other comments about `RangeLock` as to why having a `Option<WaitQueue>` field is a bad idea.